### ISSUE TITLE
Remove tensorflow from requirements.txt

### ISSUE THREAD

**Issue by blazekadam** [Cognexa/cxflow-tensorflow#3]
It is not a good practice to install tensorflow from pip by default. We should remove it from requirements.txt to encourage mindful installation/compilation suitable for the particular case.

**Comment by FloopCZ** [comment]
Are you sure about this? Imagine a common Ubuntu Joe, who just wants to train his simple network  made during his deep learning course at school. I think he really does not want to build optimized tensorflow by himself, since it is a rather long and sometimes painful process.

Furthermore, for a power user the requirement is not a problem, since when tensorflow is installed manually, pip will not install it from PyPI.

Maybe @petrbel can add his opinion as well.

**Comment by petrbel** [comment]
I totally agree with @FloopCZ. Reqs. should contain all python dependencies. It is up to the users to install it whatever they like.

**Comment by blazekadam** [comment]
Not sure if we aim for common Ubuntu Joes

Tensorflow installation is quite a different thing compared to other pypackages such as flask etc. The idea is to force to *read official tf instructions* (which may subject to change!) and install tf manually. In this way, one would avoid the following issues:
- installing CPU instead of GPU version (I got confused repetitively, in some cases, pip uninstall did not work!!!)
- installing wrong cuda/cudnn versions (you still need to install those manually and you need to match the version with tf, just a few hours ago I discovered that tf installed from pip is not compatible with arch CUDA/cudnn from pacman)

In other words, by using pip package, we are making choices that should be done mindfully by the user. *The current package is CPU only yet not compiled with CPU acceleration, i.e. its pretty crappy!!*.

**Comment by FloopCZ** [comment]
Then I suggest to add a warning and link to tensorflow installation guide to README.md, but keep tensorflow in requirements, because it *is* a requirement.

```
The current package is CPU only yet not compiled with CPU acceleration, i.e. its pretty crappy!!
```
They cannot compile it with CPU specific vectorized instructions if they do not know the target CPU. So it is  compiled for a pure x86_64 architecture instead. Which is kind of unfortunate.

(On the other hand, if we skip tensorflow from requirements, Python will throw an ImportError exception, which clearly states that tensorflow is not installed. It is, however, unusual to knowingly solve requirements this way.)

**Comment by petrbel** [comment]
This a misunderstanding.

1. `requirements.txt` must contain all requirements (btw pycharm warns you if they don't).
2. Installation can completely ignore requirements. It is our decision to use `requirements.txt` (and I think we should keep it that way)

> The current package is CPU only yet not compiled with CPU acceleration, i.e. its pretty crappy!!

Yup, but you can't compile it with an instruction which the user's CPU might not support.

> (On the other hand, if we skip tensorflow from requirements, Python will throw an ImportError exception, which clearly states that tensorflow is not installed. It is, however, unusual to knowingly solve requirements this way.)

Everytime I install software which does not handle its own requirements I feel very very angry (hello scikit).

I suggest keeping the reqs file as is. In addition, let's put a heavy warning that one should install it by themselves in advance.

Btw this is totally the same problem with numpy. The pip version is slow, the repository is ok - and we don't care anyway.

**Comment by blazekadam** [comment]
> Yup, but you can't compile it with an instruction which the user's CPU might not support.

> They cannot compile it with CPU specific vectorized instructions if they do not know the target CPU. So it is compiled for a pure x86_64 architecture instead. Which is kind of unfortunate.

I **do understand** what is going on and why it is like that. I just wanted to point that out.

> Everytime I install software which does not handle its own requirements I feel very very angry (hello scikit).

When installing SW, you should follow its instructions, one of them should be install this according to **their** instructions.

Our instructions are
```
pip install git+https://github.com/Cognexa/cxflow-tensorflow.git
```
which 
- installs a crap version of tensorflow (rly? wtf?)
- fails as cxflow is not in pypi (nobody cares?)

**Comment by blazekadam** [comment]
Our instructions should be 
- Install tensorflow according to LINK
- pip ...

Without having tensorflow in `requirements.txt`.

`requirements.txt` is pip specific thingy, not an ultimate list of dependencies or is it not?

**Comment by blazekadam** [comment]
```
(test) ➜  ~ pip install git+https://github.com/Cognexa/cxflow-tensorflow.git
Collecting git+https://github.com/Cognexa/cxflow-tensorflow.git
  Cloning https://github.com/Cognexa/cxflow-tensorflow.git to /tmp/pip-hxienkrz-build
Collecting numpy>=1.12.0 (from cxflow-tf==0.1)
  Downloading numpy-1.13.1-cp35-cp35m-manylinux1_x86_64.whl (16.9MB)
    100% |████████████████████████████████| 16.9MB 113kB/s 
Requirement already satisfied: pip>=9.0.0 in ./.pyenv/versions/3.5.3/envs/test/lib/python3.5/site-packages (from cxflow-tf==0.1)
Collecting tensorflow>=1.1 (from cxflow-tf==0.1)
  Downloading tensorflow-1.2.1-cp35-cp35m-manylinux1_x86_64.whl (34.5MB)
    100% |████████████████████████████████| 34.5MB 56kB/s 
Collecting typing>=3.5.3.0 (from cxflow-tf==0.1)
Collecting cxflow>=0.6 (from cxflow-tf==0.1)
  Could not find a version that satisfies the requirement cxflow>=0.6 (from cxflow-tf==0.1) (from versions: )
No matching distribution found for cxflow>=0.6 (from cxflow-tf==0.1)
```

**Comment by petrbel** [comment]
I agree with modifying the instructions as you mentioned.
Still, I strongly diasgree with removing TF from reqs. It should be a complete list of all modules you import from (as much as possible).

@FloopCZ ?

**Comment by blazekadam** [comment]
so you do not mind that installing our package installs slow CPU only tf? It may happen despite having tf installed properly as it may be an older version and pip updates it.

**Comment by petrbel** [comment]
pip doesn't update it unless `-U` specified.

And actually I don't think we have a choice. It's pretty uncomfortable to install TF and numpy (maybe pandas as well?) separatelly (btw differens OSs have different package names etc). This is a giant overhead for a user who just want to give it a shot.

Let's create a detailed tutorial in which we describe it properly. Advanced users can dig in a make everything as powerful as possible.

However, for a first-time user, let's make is as easy as possible. You don't need extra performance when trying things out or training cxMNIST.

**Comment by blazekadam** [comment]
So while tensorflow have 5 screens long installation instructions which are actually quite nice and allows you to select the most appropriate version, we make this decision for who, for BFUs? Why do we care?

**Comment by petrbel** [comment]
I don't :) I just specify the requirements. Let the user decide what is the approproate way of TF installation. Advanced users have optimized TF installed anyway so they aren't affected.

The newbies, let's say students attending Deep Learning course, who don't even know what TF is can experiment with it right away. Btw that's what you wanted, isn't it? Growing the community.

**Comment by blazekadam** [comment]
**You need not to specify the exhaustive list of dependencies. You need to specify the installation procedure that is sound and complete.** `requirements.txt` is just for pip. If we do not want to install a dependency through pip then it should not be included there.

**Comment by blazekadam** [comment]
So the question is: *Do we want to install tensorflow through pip by default*.

I do not.

**Comment by petrbel** [comment]
Of course, an AUR package would be much nicer (@FloopCZ triggered?). Anyway, pip is a standard in python world, so (if not by default) we should support it.

I still feel you mix requirements and installation. These are two completely independent things. We state requirements. The fact that we use them during the process of installation is something else. The user can of course install the package however they like.

I can't see any drawbacks of this approach (maybe we should update our tutorials).

**Comment by FloopCZ** [comment]
> installs a crap version of tensorflow (rly? wtf? you really do not mind?)

Well, no matter how you install TF, unless you actually compile it, it will not have the CPU-specific instructions. And the official tutorial will probably guide you to install their wheel, which is the same wheel as from ty PyPI. Compilation is for power users.

> Btw this is totally the same problem with numpy. The pip version is slow, the repository is ok - and we don't care anyway.

Not really, not even NumPy from repository is actually that much optimized. For that, you would have to link it against MKL, ATLAS, or similar. (and usually build it yourself, since repository versions do not usually do that). The same stands for repository versions of tensorflow, scipy etc.

It is a usual practice to list numpy in requirements, even though an optimized super fast version is better. However, most users just don't care about speed when they use python. And if they do, they will build their own TF and their own NumPy. I would keep the requirements as they are, since it is a standard. In the installation guilde, I would add info for those seeking for performance.

**Comment by blazekadam** [comment]
It's not a bug. It's a feature! 

Closing.
